parser.y

```diff
@@ -420,19 +424,24 @@ decl: declaration_specifiers init_declarator_list_opt SEMICOLON {
       }
     } else {
       auto decl = std::move(std::get<std::unique_ptr<DeclNode>>(decl_specifiers));
-      // A record declaration that doesn't declare any identifier, e.g., `struct point {int x, int y};`.
-      if (init_decl_list.empty()) {
-        decl_list.push_back(std::move(decl));
-      } else {
-        auto& rec_decl = dynamic_cast<RecordDeclNode&>(*decl);
-        // Initialize record variable.
+      auto& rec_decl = dynamic_cast<RecordDeclNode&>(*decl);
+
+      if (!init_decl_list.empty()) {
         for (auto& init_decl : init_decl_list) {
           if (init_decl) {
             init_decl->type = ResolveType(rec_decl.type->Clone(), std::move(init_decl->type));
           }
           decl_list.push_back(std::move(init_decl));
         }
       }
+
+      auto rec_type_id = dynamic_cast<RecordType&>(*rec_decl.type).id();
+      // TODO: handle cases where struct and union variables have the same name.
+      // Insert RecordDeclNode if it is a newly create type.
+      if (auto existing_type = custom_types.find(rec_type_id); existing_type == custom_types.end()) {
+        decl_list.insert(decl_list.begin(), std::move(decl));
```

Comment:
Is there any special reason that the declarations have to be inserted into the beginning?

Reply:
The reason is to dump type declarations in front of variable declarations.

```
          RecordDeclNode <46:10> struct book definition
            FieldNode <47:9> fiction: int
            FieldNode <48:9> sci_fi: int
            FieldNode <49:9> history: int
          RecordVarDeclNode <50:5> library: struct book
            InitExprNode <51:5> int
              IdDesNode <51:6> fiction
              IntConstExprNode <51:16> 100: int
            InitExprNode <51:5> int
              IdDesNode <52:6> sci_fi
              IntConstExprNode <52:15> 50: int
            InitExprNode <51:5> int
              IdDesNode <53:6> history
              IntConstExprNode <53:16> 500: int
```

And since we need `decl` to resolve type , I decided to insert type declaration after resolving type.

Reply:
Could you add a comment for this? :raising_hand: